File: Itido-BMS/dart/bms_dart/lib/src/repositories/task_completed_repository.dart
import 'dart:async';
import '../models/task_completed.dart';
import 'source.dart';

abstract class TaskCompletedSource extends Source {
  Future<TaskCompleted> fetch(int taskCompletedId);
  Future<List<TaskCompleted>> fetchOfTask(int taskId);
  Future<int> createFromComment(int taskId, String comment);
  Future<int> create(int taskId, TaskCompleted taskCompleted);
  Future<bool> update(int taskCompletedId, TaskCompleted taskCompleted);
}

class TaskCompletedRepository extends TaskCompletedSource {
  final List<TaskCompletedSource> sources;
  TaskCompletedRepository(this.sources);

  Future<TaskCompleted> fetch(int taskCompletedId) async {
    var values;
    for (var source in sources) {
      values = await source.fetch(taskCompletedId);
      if (values != null) {
        break;
      }
    }
    return values;
  }

  Future<List<TaskCompleted>> fetchOfTask(int taskId) async {
    var values;
    for (var source in sources) {
      values = await source.fetchOfTask(taskId);
      if (values != null) {
        break;
      }
    }
    return values;
  }

  Future<int> createFromComment(int taskId, String comment) =>
      sources[0].createFromComment(taskId, comment);

  Future<int> create(int taskId, TaskCompleted taskCompleted) =>
      sources[0].create(taskId, taskCompleted);

  Future<bool> update(int taskCompletedId, TaskCompleted taskCompleted) =>
      sources[0].update(taskCompletedId, taskCompleted);

  @override
  void dispose() {
    sources.forEach((source) => source?.dispose());
  }
}

File: Itido-BMS/RentAPI/Rent/Repositories/TaskCompletedRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Rent.ContextPoint.Exceptions;
using Rent.Data;
using Rent.Models;
using Rent.Repositories.TimePlanning;

namespace Rent.Repositories
{
    public class TaskCompletedRepository
    {
        private readonly RentContext _context;
        private readonly IRoleAuthenticationRepository _roleRepo;

        public TaskCompletedRepository(RentContext context, IRoleAuthenticationRepository roleRepo)
        {
            _context = context;
            _roleRepo = roleRepo;
        }

        public async Task<int> Create(int requester, int taskId, string comment)
        {
            return await Create(requester, taskId, new CleaningTaskCompleted
            {
                CompletedDate = Helpers.DateTimeHelpers.GmtPlusOneDateTime(),
                Comment = comment,
            });

        }

        public async Task<int> Create(int requester, int taskId, CleaningTaskCompleted taskCompleted)
        {
            if (_roleRepo.IsAdminOrManager(requester) || _roleRepo.IsUser(requester))
            {
                taskCompleted.CleaningTaskID = taskId;
                taskCompleted.CompletedByUserID = requester;
                taskCompleted.Confirmed = true;
                _context.CleaningTaskCompleted.Add(taskCompleted);
                await _context.SaveChangesAsync();
                return taskCompleted.ID;
            }
            throw new UnauthorizedAccessException();
        }

        public async Task Update(int requester, int taskCompletedId, CleaningTaskCompleted taskCompleted)
        {
            if (_roleRepo.IsAdminOrManager(requester))
            {
                var taskCompletedToUpdate = _context.CleaningTaskCompleted.Find(taskCompletedId);
                if (taskCompletedToUpdate == null) throw new NotFoundException();

                taskCompletedToUpdate.Comment = taskCompleted.Comment;
                taskCompletedToUpdate.CompletedDate = taskCompleted.CompletedDate;

                _context.CleaningTaskCompleted.Update(taskCompletedToUpdate);
                await _context.SaveChangesAsync();
            }
            else
            {
                throw new UnauthorizedAccessException();
            }
        }

        public dynamic GetDTO(int requester, int taskCompletedId)
        {
            var requesterRole = _roleRepo.GetRole(requester);
            var taskCompleted = _context.CleaningTaskCompleted.Find(taskCompletedId);
            return CleaningTaskCompleted.BasicDTO(requester, requesterRole).Compile()(taskCompleted);
        }

        public IEnumerable<dynamic> GetOfTaskDTO(int requester, int taskId)
        {
            var requesterRole = _roleRepo.GetRole(requester);
            return _context.CleaningTaskCompleted.Where(ctc => ctc.CleaningTaskID == taskId)
                .Include(ctc => ctc.CompletedByUser).Select(CleaningTaskCompleted.BasicDTO(requester, requesterRole));
        }
    }
}
